[uss_qualifier] expand the authentication validation scenario to include constraint reference endpoints

Shastick opened the pull request:
This is heavily inspired from the existing validation logic for operational intent references.

Some additional updates:
 - scopes for any endpoint group may be missing without the whole scenario being skipped
 - if all scopes are missing (ie, impossible to test any group) then a MissingResourceError is raised (which should cause the scenario to be skipped entirely)
 - added a validation step to the `utm_implementation_us` configuration to make the expected number of skipped scenarios explicit, which should allow us to detect if scenarios suddenly start being skipped when we don't expect it.

Note: this also moves the instantiations of the authentication validator classes just before each validator is run, as they depend on the current time.

BenjaminPelletier commented on monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md:
```diff
@@ -20,10 +20,11 @@ This scenario will check for the scope's availability and transparently ignore c
 Required scopes for running this scenario:
 
 - `utm.strategic_coordination`
+- `utm.constraint_management`
+- `utm.constraint_processing`
```
Making these required makes this scenario unusable in jurisdictions that don't use these features (which is all of them I'm aware of right now).  Probably the most important CI configuration currently is utm_implementation_us as it parallels a test configuration in active use for actual regulatory compliance in the US, and this causes the entire scenario to be skipped in that configuration.  We should probably [add](https://github.com/interuss/monitoring/blob/cac12985971f32765f87acedf19108bf105d283a/monitoring/uss_qualifier/configurations/dev/utm_implementation_us_baseline.libsonnet#L357) a [validation](https://github.com/interuss/monitoring/blob/cac12985971f32765f87acedf19108bf105d283a/monitoring/uss_qualifier/reports/validation/definitions.py#L161) criterion to that configuration (and probably others) that the [count](https://github.com/interuss/monitoring/blob/cac12985971f32765f87acedf19108bf105d283a/monitoring/uss_qualifier/reports/validation/definitions.py#L111) of [skipped actions](https://github.com/interuss/monitoring/blob/cac12985971f32765f87acedf19108bf105d283a/monitoring/uss_qualifier/reports/validation/definitions.py#L79) doesn't exceed the number of actions we expect to be skipped -- that would catch newly-skipped scenarios like this.

Each of these scopes should be optional (and then the validation for them not performed if not provided), or else split this scenario into strategic coordination (existing), constraint management, and constraint processing.

Shastick replied:
Ah yes, that's an oversight, thank you for pointing it out. I'll: 

- [ ] make the relevant scopes (`utm.constraint_management`, `utm.constraint_processing`) optional and make sure `utm_implementation_us` properly runs
- [ ] possibly make the required scope for any endpoint group optional
- [ ] add a validation criterion to `utm_implementation_us`

BenjaminPelletier commented on monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md:
```diff
@@ -317,6 +318,146 @@ it is in violation of **[astm.f3548.v21.DSS0210,A2-7-2,7](../../../../../require
 If the DSS does not allow searching for operational intents when valid credentials are presented,
 it is in violation of **[astm.f3548.v21.DSS0005,1](../../../../../requirements/astm/f3548/v21.md)**.
 
+### Constraint reference endpoints authentication test step
```
Probably not the highest priority for adjustment, but It seems like the "steps" in this scenario [should actually be](https://github.com/interuss/monitoring/tree/main/monitoring/uss_qualifier/scenarios#structure) "cases".  A step is supposed to be a single logical action (do one thing), reflecting what a manual checkout test card would look like ("(step 1) Ok, now we're going to X -- did things Q and P happen?  (step 2) Now we're going Y -- did things R and S happen?").  I would imagine a good test case would be "test authentication of all the constraint management endpoints", and then each testing of each constraint management endpoint would be a test step (or, testing of each constraint management endpoint in a particular way would be a test step).

Shastick replied:
This makes sense. it might be possible to easily do this as part of this PR, as I need to handle the optional `utm.constraint_processing` scopes, but I'm not entirely sure yet.

I created #742 to keep track of it in any case.

Shastick replied:
Given the current priorities I'll leave the splitting into separate test cases for later.

Shastick commented:
Updated the PR: the various endpoint groups are now only optionally tested when the relevant scopes are available.

The scenario now runs as part of the `utm_implementation_us` configuration and skips the constraints references endpoints (I also added a validation step to the relevant configuration)

```
2024-08-07 07:23:07.532 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: missing_scope -> Missing scope testing disabled.
2024-08-07 07:23:07.532 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: scd -> Testing Strategic Coordination endpoints (Subscriptions and OIRs)
2024-08-07 07:23:07.532 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: scd_wrong_scope -> Incorrect scope testing enabled with scope interuss.flight_planning.direct_automated_test.
2024-08-07 07:23:07.763 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: constraints -> Skipping Constraint Reference endpoints
2024-08-07 07:23:07.763 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: availability -> Testing Availability endpoints
2024-08-07 07:23:07.763 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: availability_wrong_scope -> Incorrect scope testing enabled for availability endpoints with scope interuss.flight_planning.direct_automated_test.
```

Shastick commented:
@BenjaminPelletier this PR ends up mixing two things now: if it makes the review too cumbersome let me know and I'll split it.

Shastick commented:
Ready for review: locally, using the f3548_self_contained configuration, all endpoints are tested:

```
2024-08-13 07:06:38.056 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: missing_scope -> Missing scope testing enabled.
2024-08-13 07:06:38.056 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: scd -> Testing Strategic Coordination endpoints (Subscriptions and OIRs)
2024-08-13 07:06:38.056 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: scd_wrong_scope -> Incorrect scope testing enabled with scope interuss.flight_planning.direct_automated_test.
2024-08-13 07:06:38.357 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: availability -> Testing Availability endpoints
2024-08-13 07:06:38.358 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: availability_wrong_scope -> Incorrect scope testing enabled for availability endpoints with scope interuss.flight_planning.direct_automated_test.
2024-08-13 07:06:38.402 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: constraints -> Testing Constraint Reference endpoints
2024-08-13 07:06:38.402 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:316 - Note: constraints_wrong_scope -> Incorrect scope testing enabled with scope interuss.flight_planning.direct_automated_test.
```

mickmis commented on monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/cr_api_validator.py:
```diff
@@ -0,0 +1,559 @@
+                abs(
+                    cr.time_end.value.datetime
+                    - self._current_cr.time_end.value.datetime
+                ).total_seconds()
```
Why not just checking the version instead of comparing the end time? If an update is made for which the end time remains the same, this check won't catch it.

Shastick replied:
I might have been wondering if the version might not get updated in certain circumstances and decided to check one of the parameters that got mutated.

It costs nothing to also check for the version so I'll add that.

mickmis reviewed (commented):
Some minor comments, otherwise LGTM.

Shastick commented on monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md:
```diff
@@ -387,6 +388,146 @@ it is in violation of **[astm.f3548.v21.DSS0100,1](../../../../../requirements/a
+#### 🛑 Create constraint reference with valid credentials check
+
+If the DSS does not allow the creation of a constraint reference when valid credentials are presented,
+it is in violation of **[astm.f3548.v21.DSS0005,1](../../../../../requirements/astm/f3548/v21.md)**.
```
Good catch!

mickmis replied:
Still 1?

Shastick commented on monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md:
```diff
@@ -387,6 +388,146 @@ it is in violation of **[astm.f3548.v21.DSS0100,1](../../../../../requirements/a
+#### 🛑 Get constraint reference with valid credentials check
+
+If the DSS does not allow fetching a constraint reference when valid credentials are presented,
+it is in violation of **[astm.f3548.v21.DSS0005,1](../../../../../requirements/astm/f3548/v21.md)**.
```
Good catch. Added to the validator.

mickmis replied:
Still 1?

mickmis commented on monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md:
```diff
@@ -387,6 +388,146 @@ it is in violation of **[astm.f3548.v21.DSS0100,1](../../../../../requirements/a
+#### 🛑 Delete constraint reference with valid credentials check
+
+If the DSS does not allow the deletion of a constraint reference when valid credentials are presented,
+it is in violation of **[astm.f3548.v21.DSS0005,1](../../../../../requirements/astm/f3548/v21.md)**.
```
Still 1?

mickmis reviewed (approved):
Could you fix the missed requirements? Otherwise LGTM thanks!

Shastick commented:
> Could you fix the missed requirements? Otherwise LGTM thanks!

Thanks for catching them, they did not make it into the previous commit. Should be correct now.
